File: python/day02.py
from math import prod

def get_counts(pull):
    return {name: int(count) for count, name in [comp.split(" ") for comp in pull.split(", ")]} 
        
def max_pull(pulls):
    m = {"red":0, "green":0, "blue":0}
    for c in pulls:
        for key in c:
            m[key] = max(m[key], c[key])
    return m

limits = {"red":12, "green":13, "blue":14}
i = 0
total = 0
power = 0
while True:
    try:
        i += 1
        line = input().split(": ")[1]
        m = max_pull(get_counts(turn) for turn in line.split("; "))
        if all(m[k] <= limits[k] for k in limits):
            total += i
        power += prod(m.values())
    except:
        print(total)
        print(power)
        break

File: haskell/src/Day02.hs
module Day02 (run) where
 
import Lib

data Count = Count Int Int Int deriving Show

instance Semigroup Count where
  (<>) (Count a b c) (Count x y z) = Count (max a x) (max b y) (max c z)
   
instance Monoid Count where
  mempty = Count 0 0 0
  
readCount :: String -> Count
readCount s = case words s of
  i:"red":_ -> Count (read i) 0 0
  i:"green":_ -> Count 0 (read i) 0
  i:"blue":_ -> Count 0 0 (read i)
  _ -> Count 0 0 0
  
score :: Num a => a -> Count -> a
score i (Count a b c) = if a <= 12 && b <= 13 && c <= 14 then i else 0

power :: Count -> Int
power (Count a b c) = a * b * c

maxPull :: String -> Count
maxPull = foldMap (foldMap readCount . split ',') . split ';'

matchingTotal :: [String] -> Integer
matchingTotal = sum . zipWith score [1..] . map maxPull

totalPower :: [String] -> Int
totalPower = sum . map (power . maxPull)

getPulls :: String -> String
getPulls line = split ':' line !! 1

run :: IO String -> IO ()
run s = do
  print . matchingTotal . map getPulls . lines =<< s
  print . totalPower . map getPulls . lines =<< s